File: heeyeon/9주차/134. Gas Station.java
//시간 줄이는 방법 생각해 보기
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int remain = 0;
        int idx = 0;
        int cnt = 0;
        
        //처음에 가능한지 체크
        remain = gas[idx] - cost[idx];
        while(remain >= 0){
            idx++;
            cnt++;
            remain += (gas[idx%gas.length] - cost[idx%gas.length]);
            if(cnt == gas.length && remain >= 0)
                return 0;
        }
        
        for(int i=0; i<gas.length-1; i++){
            //시작 없을때 다음 시작점에서 가능한지 확인
            if(gas[i] < cost[i]){
                idx = i+1;
                cnt = 0;
                remain = gas[idx] - cost[idx];
                while(remain >= 0){
                    idx++;
                    cnt++;
                    remain += (gas[idx%gas.length] - cost[idx%gas.length]);
                    if(cnt == gas.length && remain >= 0)
                        return i+1;
                }
            }
        }
        return -1;
    }
}

File: taewan/week 9/134. Gas Station.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        gc = [gas[i] - cost[i] for i in range(len(gas))]
        gas, start, n = 0, 0, len(gc)
        for i in range(n):
            if gas + gc[i] >= 0:
                gas += gc[i]
                continue
            gas = 0
            start = i + 1
        
        gas = 0
        for i in range(n):
            gas += gc[(i + start) % n]
            if gas < 0:
                return -1
        
        return start